extensions/amp-web-push/0.1/worker-messenger.js

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+class WorkerMessengerReplyBuffer {
+  constructor() {
+    this.replies = {};
+  }
+
+  addListener(command, callback, onceListenerOnly) {
+    const record = {
+      callback: callback,
+      onceListenerOnly: onceListenerOnly
+    };
+
+    if (this.findListenersForMessage(command).length > 0) {
+      this.replies[command.toString()].push(record);
+    } else {
+      this.replies[command.toString()] = [record];
+    }
+  }
+
+ findListenersForMessage(command) {
+    return this.replies[command.toString()] || [];
+  }
+
+  deleteListenerRecords(command) {
+    this.replies[command.toString()] = null;
+  }
+
+  deleteListenerRecord(command, targetRecord) {
+    const listenersForCommand = this.replies[command.toString()];
+    for (
+      let listenerRecordIndex = listenersForCommand.length - 1;
+      listenerRecordIndex >= 0;
+      listenerRecordIndex--) {
+      const listenerRecord = listenersForCommand[listenerRecordIndex];
+      if (listenerRecord === targetRecord) {
+        listenersForCommand.splice(listenerRecordIndex, 1);
+      }
+    }
+  }
+}
+
+class WorkerMessenger {
```

Comment:
I decided to [remove `worker-messenger.js`](https://github.com/ampproject/amphtml/pull/10468/commits/47a45270558ac18d7d2c53ae5551209eb92be2a6) in the end.

The bulk of the code was around the abstraction of providing a `once()` listener, so that you could do:

`workerMessenger.once('some-topic', myCallback())`

I was able to find a reasonable workaround for it. I took out the worker messenger from the extension + the sample file that used to use it.